tsdb/head_read.go

```diff
@@ -658,8 +698,14 @@ func (s *memSeries) iterator(id chunks.HeadChunkID, c chunkenc.Chunk, isoState *
 			}
 		}
 
-		if s.headChunk != nil {
-			totalSamples += s.headChunk.chunk.NumSamples()
+		ix -= len(s.mmappedChunks)
+		if s.headChunks != nil {
+			for j, elem := range s.headChunks.toReversedSlice() {
+				totalSamples += elem.chunk.NumSamples()
+				if j < ix {
+					previousSamples += elem.chunk.NumSamples()
+				}
+			}
```

Comment:
Since this iterator is called for every chunk of every series in a query, we can avoid a lot of slice allocation by just getting the `l := s.headChunks.len()`, and do a simple linkedlist traversal and use the `l` and some index to update the `previoudSamples`.